deps/3rdparty/libdispatch/src/queue.c

```diff
@@ -1978,3 +2019,92 @@ dispatch_after_f(dispatch_time_t when, dispatch_queue_t queue, void *ctxt, void
 	dispatch_resume(as_do(ds));
 }
 
+struct specificContainer {
+	dispatch_queue_t operationQueue;
+	void* context;
+};
+
+static void _dispatch_queue_init_specific_list(dispatch_queue_t dq) {
+	dq->dq_specific_q = dispatch_queue_create("SpecificQueue", NULL);
+	dispatch_queue_specific_list_t dqsl;
+	dqsl = calloc(1, sizeof(struct dispatch_queue_specific_list_s));
+	TAILQ_INIT(&dqsl->contextList);
+	dq->dq_specific_q->dq_specific_list = dqsl;
+}
+
+static void _dispatch_queue_insert_specific(dispatch_queue_specific_t dqs) {
+	dispatch_queue_t dq = _dispatch_queue_get_current();
+	dispatch_queue_specific_list_t dqsl = dq->dq_specific_list;
+	dispatch_queue_specific_t var;
+
+	TAILQ_FOREACH(var, &dqsl->contextList, specific) {
+		if (var->key == dqs->key) {
+			if(var->destructor) {
+				var->destructor(var->context);
+			}
+
+			// If a specific exists with this key, simply update the context on that specific
+			if(dqs->context) {
+				var->context = dqs->context;
+				var->destructor = dqs->destructor;
+			} else {
+				TAILQ_REMOVE(&dqsl->contextList, var, specific);
+				free(var);
+			}
+			free(dqs);
+			return;
+		}
+	}
+
+	if(dqs->context) {
+		TAILQ_INSERT_TAIL(&dqsl->contextList, dqs, specific);
+	}
+}
+
+void dispatch_queue_set_specific(dispatch_queue_t queue, const void *key, void *_Nullable context, dispatch_function_t _Nullable destructor) {
+	if (slowpath(!key)) {
+		return;
+	}
+	dispatch_queue_specific_t specific = calloc(1, sizeof(struct dispatch_queue_specific_s));
+	specific->key = key;
+	specific->context = context;
+	specific->destructor = destructor;
+
+	// Delayed initialization
+	if (slowpath(!queue->dq_specific_q)) {
+		_dispatch_queue_init_specific_list(queue);
+	}
+
+	dispatch_barrier_sync_f(queue->dq_specific_q, specific, _dispatch_queue_insert_specific);
+}
+
+static void _dispatch_queue_get_specific(void* context) {
+	dispatch_queue_t dq = _dispatch_queue_get_current();
+	dispatch_queue_specific_list_t dqsl = dq->dq_specific_list;
+	dispatch_queue_specific_t var;
+
+	// The parameter passed in is reference to a void* key
+	void** ctxt = context;
```

Comment:
you can save this temporary by just specifying that `context` is itself a `void**`

Reply:
We use it as a void* out pointer though for the return value of the context.